pkg/attachutils/simpleattach.go

```diff
@@ -0,0 +1,385 @@
+			if stdout != nil {
+				conn.Close()
+			}
+			if stderr != nil {
+				conn.Close()
+			}
+		}
+		log.Debugf("attach: stdin: end")
+		wg.Done()
+	}()
+
+	attachStream := func(name string, stream io.Writer) {
+		if stream == nil {
+			return
+		}
+
+		log.Debugf("attach: %s: begin", name)
+		_, err := io.Copy(stream, bufReader)
+		if err == io.ErrClosedPipe {
+			err = nil
+		}
+		if err != nil {
+			log.Errorf("attach: %s: %v", name, err)
+			errors <- err
+		}
+		// Make sure stdin gets closed
+		if stdin != nil {
+			stdin.Close()
+		}
+		conn.Close()
+		log.Debugf("attach: %s: end", name)
+		wg.Done()
+	}
+
+	go attachStream("stdout", stdout)
+	go attachStream("stderr", stderr)
+
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+	<-done
+	close(errors)
+	for err := range errors {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Code copied from Docker and modified.  client streams belong to the hijacked
+// connection and server streams belong to the tether ssh server
+func AttachStreams(tty bool, clientStdin io.ReadCloser, clientStdout, clientStderr io.Writer, svrStdin io.WriteCloser, svrStdout io.Reader, svrStderr io.Reader, keys []byte) error {
+	var (
+		wg     sync.WaitGroup
+		errors = make(chan error, 3)
+	)
+
+	if clientStdin != nil {
+		wg.Add(1)
+	}
+
+	if clientStdout != nil {
+		wg.Add(1)
+	}
+
+	if clientStderr != nil {
+		wg.Add(1)
+	}
+
+	// Connect stdin of container to the http conn.
+	go func() {
+		if clientStdin == nil {
+			return
+		}
+		log.Debugf("attach: stdin: begin")
+
+		var err error
+		if tty {
+			_, err = copyEscapable(svrStdin, clientStdin, keys)
+		} else {
+			_, err = io.Copy(svrStdin, clientStdin)
+		}
+		if err == io.ErrClosedPipe {
+			err = nil
+		}
+		if err != nil {
+			log.Printf("attach: stdin: %s\n", err)
+			errors <- err
+		}
+		if !tty {
+			svrStdin.Close()
+		}
+		log.Debugf("attach: stdin: end")
+		wg.Done()
+	}()
+
+	attachStream := func(name string, clientWriter io.Writer, svrReader io.Reader) {
+		log.Debugf("attach: %s: begin", name)
+		_, err := io.Copy(clientWriter, svrReader)
+		if err == io.ErrClosedPipe {
+			err = nil
+		}
+		if err != nil {
+			log.Errorf("attach: %s: %v", name, err)
+			errors <- err
+		}
+		// Make sure stdin gets closed
+		if clientStdin != nil {
+			clientStdin.Close()
+		}
+		log.Debugf("attach: %s: end", name)
+		wg.Done()
+	}
+
+	go attachStream("stdout", clientStdout, svrStdout)
+	go attachStream("stderr", clientStderr, svrStderr)
+
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+	<-done
+	close(errors)
+	for err := range errors {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Code c/c from io.Copy() modified to handle escape sequence
+func copyEscapable(dst io.Writer, src io.ReadCloser, keys []byte) (written int64, err error) {
+	if len(keys) == 0 {
+		// Default keys : ctrl-p ctrl-q
+		keys = []byte{16, 17}
```

Comment:
The only reason this function exists in the docker code is to handle this condition.  In our case, we can do better.  If we were to create a REST endpoint on the PL which handles "detach", then we simply need to capture the cntrl-p cntrl-q in the docker personality and call the REST detach to close the connection without closing `stdin`, `stdout`, `stderr` on the process in the container.